### PR TITLE
fix(rust, python): Clear window cache after evaluate predication expr

### LINKED ISSUE #10499: Multiple filters in LazyFrame on same column throws shape error
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
df = pl.from_dict(
    {"a": [None, None, 3, None, 5, 0, 0, 0, 9, 10], "b": [1, 1, 2, 2, 3, 3, 4, 4, 5, 5]}
)

df.lazy().filter((pl.col("a").null_count() < pl.count()).over("b")).filter(
    ((pl.col("a") == 0).sum() < pl.count()).over("b")
).collect()
```

### Issue description

Doing the above causes an error that the second filter has the wrong shape. It is comparing the full DataFrame of length 10 to the filtered shape of 8. 
`ShapeError: filter's length: 10 differs from that of the series: 8`

This same error does not happen if you collect in between the 2 filter calls.

### Expected behavior

I would expect for the DataFrame to filter out both where `b==1` and `b==4` without throwing any errors. 

### Installed versions

<details>

```
--------Version info---------
Polars:              0.18.15
Index type:          UInt32
Platform:            macOS-13.3.1-arm64-arm-64bit
Python:              3.11.3 (main, Jul 14 2023, 09:34:26) [Clang 14.0.3 (clang-1403.0.22.14.1)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         2.2.1
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              2023.5.0
matplotlib:          3.7.2
numpy:               1.24.3
pandas:              1.5.3
pyarrow:             10.0.1
pydantic:            1.10.12
sqlalchemy:          2.0.15
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by reswqa**
This fixes #10499.

### CODE REVIEW

File: crates/polars-lazy/src/physical_plan/executors/filter.rs [hunk 1]
```diff
@@ -35,6 +35,9 @@ impl Executor for FilterExec {
             state.insert_has_window_function_flag()
         }
         let s = self.predicate.evaluate(&df, state)?;
+        if state.cache_window() {
```
Comment (reswqa) [inline crates/polars-lazy/src/physical_plan/executors/filter.rs:38]: Might replace this by `if has_window` to reduce the overhead of atomic load? 🤔

Comment (ritchie46) [reply crates/polars-lazy/src/physical_plan/executors/filter.rs:38]: Yes an you put the `cache_window` and the `clear_window_expr` behind a `if self.has_windows`?

File: crates/polars-lazy/src/physical_plan/executors/scan/mod.rs [hunk 2]
```diff
@@ -114,6 +117,9 @@ impl Executor for AnonymousScanExec {
                 (false, Some(predicate)) => {
                     let mut df = self.function.scan(self.options.clone())?;
                     let s = predicate.evaluate(&df, state)?;
+                    if state.cache_window() {
```
Comment (ritchie46) [inline crates/polars-lazy/src/physical_plan/executors/scan/mod.rs:120]: This should always evaluate to `false` if this executor didn't activate cached window functions. 

Can we add a `has_window` boolean similar to in `FilterExec` and recreate that behavior?

File: crates/polars-lazy/src/physical_plan/executors/scan/mod.rs [hunk 3]
```diff
@@ -84,6 +84,9 @@ impl Executor for DataFrameExec {
                 state.insert_has_window_function_flag()
             }
             let s = selection.evaluate(&df, state)?;
+            if state.cache_window() {
```
Comment (ritchie46) [inline crates/polars-lazy/src/physical_plan/executors/scan/mod.rs:87]: See comments in AnonymousScan

**Review by ritchie46 — COMMENTED** [review]
Nice @reswqa. Great analysis.

I have left some comments.

File: crates/polars-lazy/src/physical_plan/executors/filter.rs (hunk 1 above)
Comment (reswqa) [reply crates/polars-lazy/src/physical_plan/executors/filter.rs:38]: Do you mean to replace it with

```rust
if state.has_window() && state.cache_window() {
   state.clear_window_expr_cache()
}
```

File: crates/polars-lazy/src/physical_plan/executors/scan/mod.rs (hunk 3 above)
Comment (reswqa) [reply crates/polars-lazy/src/physical_plan/executors/scan/mod.rs]: It seems that `DataStreamScan` already has `predicate_has_windows` field.

File: crates/polars-lazy/src/physical_plan/executors/scan/mod.rs [hunk 4]
```diff
@@ -101,29 +104,41 @@ pub(crate) struct AnonymousScanExec {
...
     pub(crate) predicate: Option<Arc<dyn PhysicalExpr>>,
+    pub(crate) predicate_has_windows: bool,
 }
 
 impl Executor for AnonymousScanExec {
     fn execute(&mut self, state: &mut ExecutionState) -> PolarsResult<DataFrame> {
-        state.record(
-            || match (self.function.allows_predicate_pushdown(), &self.predicate) {
-                (true, Some(predicate)) => {
+        match (self.function.allows_predicate_pushdown(), &self.predicate) {
```
Comment (reswqa) [inline crates/polars-lazy/src/physical_plan/executors/scan/mod.rs:112]: This change is to make the borrow checker happy (due to the mutable borrow of `state`). Is there a simpler approach here?

**Comment by reswqa** [comment]
Thanks @ritchie46 for the review. I have updated this, but I'm not sure if I totally understand all comments, feel free to append new comments.

File: crates/polars-lazy/src/physical_plan/executors/filter.rs (hunk 1 above)
Comment (ritchie46) [reply crates/polars-lazy/src/physical_plan/executors/filter.rs]: This is what we do during projection.

```rust
    if has_windows {
        state.clear_window_expr_cache();
    }
```

File: crates/polars-lazy/src/physical_plan/executors/scan/mod.rs (hunk 3 above)
Comment (ritchie46) [reply crates/polars-lazy/src/physical_plan/executors/scan/mod.rs]: Oh great. Then we can use that. :)

File: crates/polars-lazy/src/physical_plan/executors/scan/mod.rs (hunk 4 above)
Comment (ritchie46) [reply crates/polars-lazy/src/physical_plan/executors/scan/mod.rs:112]: Can take a look later. I think it is fine for this PR.

File: crates/polars-lazy/src/physical_plan/executors/filter.rs (hunk 1 above)
Comment (reswqa) [reply crates/polars-lazy/src/physical_plan/executors/filter.rs]: I see, removed it.

**Comment by ritchie46** [comment]
Alright! :raised_hands:

**Pull request merged**